ee/clickhouse/views/person.py

```diff
@@ -1,30 +1,93 @@
-from typing import Callable, Dict, Optional, Tuple
+import json
+from typing import Any, Callable, Dict, List, Optional, Tuple, cast
 
+from django.db.models.query import Prefetch, QuerySet
 from rest_framework.decorators import action
 from rest_framework.exceptions import NotFound
 from rest_framework.request import Request
 from rest_framework.response import Response
 
 from ee.clickhouse.client import sync_execute
 from ee.clickhouse.models.person import delete_person
+from ee.clickhouse.models.property import parse_prop_clauses
 from ee.clickhouse.queries.clickhouse_retention import ClickhouseRetention
 from ee.clickhouse.queries.clickhouse_stickiness import ClickhouseStickiness
 from ee.clickhouse.queries.funnels import ClickhouseFunnelPersons, ClickhouseFunnelTrendsPersons
 from ee.clickhouse.queries.trends.lifecycle import ClickhouseLifecycle
 from ee.clickhouse.sql.person import GET_PERSON_PROPERTIES_COUNT
 from posthog.api.person import PersonViewSet
-from posthog.api.utils import format_next_url, format_offset_absolute_url
-from posthog.constants import INSIGHT_FUNNELS, FunnelVizType
+from posthog.api.utils import format_offset_absolute_url
+from posthog.constants import FunnelVizType
 from posthog.decorators import cached_function
+from posthog.exceptions import UnsupportedFeature
 from posthog.models import Event, Filter, Person
 
+BASE_FILTER_PERSONS_QUERY = """
+SELECT person.id FROM person
+LEFT OUTER JOIN (
+    SELECT * FROM person_distinct_id WHERE person_distinct_id.team_id = %(team_id)s
+) person_distinct_id
+ON person.id = person_distinct_id.person_id
+WHERE team_id = %(team_id)s"""
+
+
+def filter_persons_ch(team_id: int, request: Request, queryset: QuerySet) -> QuerySet:
+    # Keep functionality in sync with posthog/api/person.py
+    params: Dict[str, Any] = {"team_id": team_id}
+    and_conditions: List[str] = []
+    if request.GET.get("id"):
+        raise UnsupportedFeature("filtering persons by field `id`")
+    if request.GET.get("uuid"):
+        params["uuids"] = request.GET["uuid"].split(",")
+        and_conditions.append("AND has(%(uuids)s, toString(person.id))")
+    if request.GET.get("search"):
+        parts = request.GET["search"].split(" ")
+        contains = []
+        for part_index, part in enumerate(parts):
```

Comment:
Worth extracting a helper to retain identical behavior with postgres' implementation?

Reply:
I feel like a helper would just be more confusing considering that the general logic is the same, but it's the bottom of each branch that looks very different between the two versions.

Reply:
If this is causing issues for you, worth making a change to e.g. `def sync_execute` instead in a separate PR? Or using mypy? :)

Reply:
Comment in the wrong place I think (should be in "Q: Why do we need the cast here?")?

Reply:
Let's take a step back.

Fundamentally this function is just doing too many things:

1. Extracting "meaning" from search query (e.g. a filter)
  2. Including some very specific parsing logic
2. Constructing query based on that
3. Lastly querying both ch and postgres based off of it

Everything we can do to extract these different pieces fundamentally makes the code easier to follow since it can make use of shared patterns.